internal/app/machined/pkg/runtime/v1alpha1/v1alpha1_events.go

```diff
@@ -5,97 +5,122 @@
 package v1alpha1
 
 import (
+	"context"
 	"fmt"
-	"log"
 	"sync"
 
 	"github.com/golang/protobuf/proto"
-	"github.com/golang/protobuf/ptypes/any"
 
-	"github.com/talos-systems/talos/api/machine"
 	"github.com/talos-systems/talos/internal/app/machined/pkg/runtime"
 )
 
 // Events represents the runtime event stream.
 type Events struct {
-	subscribers []chan machine.Event
+	// stream is used as ring buffer of events
+	stream []runtime.Event
 
-	sync.Mutex
+	// writePos is the index in streams for the next write (publish)
+	writePos int
+
+	// gen tracks number of wraparounds in stream
+	gen int64
+
+	// cap is capacity of streams
+	cap int
+
+	// mutext protects access to writePos, gen and stream
+	mu sync.Mutex
+	c  *sync.Cond
 }
 
 // NewEvents initializes and returns the v1alpha1 runtime event stream.
-func NewEvents(n int) *Events {
+func NewEvents(cap int) *Events {
 	e := &Events{
-		subscribers: make([]chan machine.Event, 0, n),
+		stream: make([]runtime.Event, cap),
+		cap:    cap,
 	}
 
+	e.c = sync.NewCond(&e.mu)
+
 	return e
 }
 
 // Watch implements the Events interface.
 func (e *Events) Watch(f runtime.WatchFunc) {
-	ch := e.add()
+	ctx, ctxCancel := context.WithCancel(context.Background())
+
+	ch := make(chan runtime.Event)
 
 	go func() {
-		defer close(ch)
-		defer e.delete(ch)
+		defer ctxCancel()
 
 		f(ch)
 	}()
-}
 
-// Publish implements the Events interface.
-func (e *Events) Publish(msg proto.Message) {
-	value, err := proto.Marshal(msg)
-	if err != nil {
-		log.Printf("failed to marshal message: %v", err)
-
-		return
-	}
-
-	event := machine.Event{
-		Data: &any.Any{
-			// In the future, we can publish `talos/runtime`, and
-			// `talos/plugin/<plugin>` (or something along those lines) events.
-			TypeUrl: fmt.Sprintf("talos/runtime/%s", proto.MessageName(msg)),
-			Value:   value,
-		},
-	}
+	e.mu.Lock()
+	pos := e.writePos
+	gen := e.gen
+	e.mu.Unlock()
 
-	e.Lock()
-	defer e.Unlock()
+	go func() {
+		defer close(ch)
 
-	for _, sub := range e.subscribers {
-		// drop the event if some subscriber is stuck
-		// dropping is bad, but better than blocking event propagation
-		select {
-		case sub <- event:
-		default:
+		for {
+			e.mu.Lock()
+			for pos == e.writePos {
```

Comment:
I'm struggling to understand when this would be false or true and why we need this.

Reply:
we start with `pos == e.writePos`, so consumer waits for the entry to be published by the publisher.

as publisher writes another entry to `e.stream`, it increases `e.writePos` (with mod, so it might go back to zero as well), and consumer picks up element from `pos` and increments `pos` until it once again hits `e.writePos`